feature_engine/creation/decision_tree_features.py

```diff
@@ -0,0 +1,422 @@
+        The random_state to initialise the training of the decision tree. It is one
+        of the parameters of the Scikit-learn's DecisionTreeRegressor() or
+        DecisionTreeClassifier(). For reproducibility it is recommended to set
+        the random_state to an integer.
+
+
+    Attributes
+    ----------
+    {variables_}
+
+    {feature_names_in_}
+
+    {n_features_in_}
+
+    Methods
+    -------
+    fit:
+        Fits the decision tree(s).
+
+    transform:
+        Adds new features.
+
+    {fit_transform}
+
+    See Also
+    --------
+    feature_engine.discretisation.DecisionTreeDiscretiser
+    feature_engine.encoding.DecistionTreeEncoder
+    sklearn.tree.DecisionTreeClassifier
+    sklearn.tree.DecisionTreeRegressor
+
+    References
+    ----------
+    .. [1] Niculescu-Mizil, et al. "Winning the KDD Cup Orange Challenge with Ensemble
+        Selection". JMLR: Workshop and Conference Proceedings 7: 23-34. KDD 2009
+        http://proceedings.mlr.press/v7/niculescu09/niculescu09.pdf
+    """
+
+    def __init__(
+        self,
+        variables: Union[None, int, str, List[Union[str, int]]] = None,
+        strategy: Union[int, List[int], Tuple[tuple, ...]] = None,
+        drop_original: bool = False,
+        cv=3,
+        scoring: str = "neg_mean_squared_error",
+        param_grid: Optional[Dict[str, Union[str, int, float, List[int]]]] = None,
+        regression: bool = True,
+        random_state: Optional[int] = None,
+    ) -> None:
+
+        if (
+                not isinstance(strategy, (int, list, tuple))
+                and strategy is not None
+        ):
+            raise ValueError(
+                f"strategy must integer, list or tuple. Got {strategy} instead."
+            )
+        # check user is not creating combinations of more variables
+        # than the number of variables provided in the 'variables' param.
+        if isinstance(strategy, int) and strategy > len(variables):
+            raise ValueError(
+                "strategy cannot be greater than the number of features to combine. "
+                f"Got strategy={strategy} and {len(variables)} features."
+            )
+
+        if isinstance(strategy, list):
+            # Check (1) user is not creating combinations of more variables
+            # than the number of variables in the 'variables' param or
+            # (2) the list is only comprised of integers.
+            if (
+                    max(strategy) > len(variables)
+                    or not all(isinstance(feature, int) for feature in strategy)
+            ):
+                raise ValueError(
+                    "strategy must be a list of integers and the maximum value cannot "
+                    "be greater than the number of variables to combine. "
+                )
+
+        if isinstance(strategy, tuple) and variables is not None:
+
+            # Check that the tuples only contains strings with variable names.
+            if (
```

Comment:
To make our lives easier, if strategy is tuple, we only allow tuples inside. Each tuple can contain one or more strings.

I edited the logic, but I might have broken somewhere accidentally. If you find that my logic breaks your amazing hard work, either try to fix it, or let me know and I'll do it :)